core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -513,13 +516,44 @@ private[spark] object Utils extends Logging {
       Files.move(sourceFile, destFile)
```

Comment:
We talked about this offline; it seems that, in Spark's case, both source and destination are on the same filesystem, so this should work.

Guava does something akin to `if (!from.renameTo(to)) copy(from, to)` internally; that works on Unix for directories if both source and destination are on the same filesystem. On Windows, though, it would fail: `renameTo` fails on Windows if the destination already exists, and the Guava code would revert back to `copy()` which only works on files. (That also means `move` is unnecessarily expensive on Windows then the destination exists, but that's a Guava issue.)

So if we care about Windows here we should probably do a `deleteRecursively` before trying the rename. It ceases being an "atomic" operation, but Java doesn't expose the Window's version of [`rename`](https://msdn.microsoft.com/en-us/library/zw5t957f.aspx), which follows the Unix semantics.